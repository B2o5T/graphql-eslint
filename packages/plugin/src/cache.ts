// Based on the `eslint-plugin-import`'s cache
// https://github.com/import-js/eslint-plugin-import/blob/main/utils/ModuleCache.js
import debugFactory from 'debug';

const log = debugFactory('graphql-eslint:ModuleCache');

export class ModuleCache<K, T> {
  map = new Map<K, { lastSeen: [number, number]; result: T }>();

  set(cacheKey: K, result: T): void {
    this.map.set(cacheKey, { lastSeen: process.hrtime(), result });
    log('setting entry for', cacheKey);
  }

  get(cacheKey: K, settings = { lifetime: 10 /* seconds */ }): T | void {
    const value = this.map.get(cacheKey);
    if (!value) {
      log('cache miss for', cacheKey);
      return;
    }
    const { lastSeen, result } = value;
    // check freshness
    if (
      process.env.NODE /* don't check for ESLint CLI */ ||
      process.hrtime(lastSeen)[0] < settings.lifetime
    ) {
      return result;
    }
  }
}
