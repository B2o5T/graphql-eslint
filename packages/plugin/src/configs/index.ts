import { ConfigName } from '../types.js';
// @ts-expect-error -- complains about no default export
import operationsAllConfig from './operations-all.js';
// @ts-expect-error -- complains about no default export
import operationsRecommendedConfig from './operations-recommended.js';
// @ts-expect-error -- complains about no default export
import schemaAllConfig from './schema-all.js';
// @ts-expect-error -- complains about no default export
import schemaRecommendedConfig from './schema-recommended.js';
// @ts-expect-error -- complains about no default export
import relayConfig from './schema-relay.js';

export const configs = {
  'schema-recommended': schemaRecommendedConfig,
  'schema-all': schemaAllConfig,
  'schema-relay': relayConfig,
  'operations-recommended': operationsRecommendedConfig,
  'operations-all': operationsAllConfig,
  'flat/schema-recommended': schemaRecommendedConfig.rules,
  'flat/schema-all': {
    ...schemaRecommendedConfig.rules,
    ...schemaAllConfig.rules,
  },
  'flat/schema-relay': relayConfig.rules,
  'flat/operations-recommended': operationsRecommendedConfig.rules,
  'flat/operations-all': {
    ...operationsRecommendedConfig.rules,
    ...operationsAllConfig.rules,
  },
} satisfies Record<ConfigName | `flat/${ConfigName}`, unknown>;
