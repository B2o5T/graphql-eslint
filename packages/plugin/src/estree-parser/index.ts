export * from "./converter";
export * from "./estree-ast";
export * from "./utils";
