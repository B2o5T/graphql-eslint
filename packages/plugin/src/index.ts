import { configs } from './configs/index.js';
import { parseForESLint, parser } from './parser.js';
import { processor } from './processor.js';
import { rules } from './rules/index.js';

export * from './types.js';
export type { IGraphQLConfig } from 'graphql-config';
export type { GraphQLTagPluckOptions } from '@graphql-tools/graphql-tag-pluck';

export { requireGraphQLSchemaFromContext, requireSiblingsOperations } from './utils.js';

export const processors = { graphql: processor };

export { rules, configs, parser, parseForESLint };

// eslint-disable-next-line import/no-default-export -- It's common practice for ESLint plugins that supports Flat config to use the default export
export default {
  parser,
  processor,
  rules,
  configs,
};
