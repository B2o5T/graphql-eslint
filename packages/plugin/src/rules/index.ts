/*
 * ðŸš¨ IMPORTANT! Do not manually modify this file. Run: `pnpm generate:configs`
 */

import { rule as alphabetize } from './alphabetize/index.js';
import { rule as descriptionStyle } from './description-style/index.js';
import { GRAPHQL_JS_VALIDATIONS } from './graphql-js-validation.js';
import { rule as inputName } from './input-name/index.js';
import { rule as loneExecutableDefinition } from './lone-executable-definition/index.js';
import { rule as matchDocumentFilename } from './match-document-filename/index.js';
import { rule as namingConvention } from './naming-convention/index.js';
import { rule as noAnonymousOperations } from './no-anonymous-operations/index.js';
import { rule as noDeprecated } from './no-deprecated/index.js';
import { rule as noDuplicateFields } from './no-duplicate-fields/index.js';
import { rule as noHashtagDescription } from './no-hashtag-description/index.js';
import { rule as noOnePlaceFragments } from './no-one-place-fragments/index.js';
import { rule as noRootType } from './no-root-type/index.js';
import { rule as noScalarResultTypeOnMutation } from './no-scalar-result-type-on-mutation/index.js';
import { rule as noTypenamePrefix } from './no-typename-prefix/index.js';
import { rule as noUnreachableTypes } from './no-unreachable-types/index.js';
import { rule as noUnusedFields } from './no-unused-fields/index.js';
import { rule as relayArguments } from './relay-arguments/index.js';
import { rule as relayConnectionTypes } from './relay-connection-types/index.js';
import { rule as relayEdgeTypes } from './relay-edge-types/index.js';
import { rule as relayPageInfo } from './relay-page-info/index.js';
import { rule as requireDeprecationDate } from './require-deprecation-date/index.js';
import { rule as requireDeprecationReason } from './require-deprecation-reason/index.js';
import { rule as requireDescription } from './require-description/index.js';
import { rule as requireFieldOfTypeQueryInMutationResult } from './require-field-of-type-query-in-mutation-result/index.js';
import { rule as requireImportFragment } from './require-import-fragment/index.js';
import { rule as requireNullableFieldsWithOneof } from './require-nullable-fields-with-oneof/index.js';
import { rule as requireNullableResultInRoot } from './require-nullable-result-in-root/index.js';
import { rule as requireSelections } from './require-selections/index.js';
import { rule as requireTypePatternWithOneof } from './require-type-pattern-with-oneof/index.js';
import { rule as selectionSetDepth } from './selection-set-depth/index.js';
import { rule as strictIdInTypes } from './strict-id-in-types/index.js';
import { rule as uniqueEnumValueNames } from './unique-enum-value-names/index.js';
import { rule as uniqueFragmentName } from './unique-fragment-name/index.js';
import { rule as uniqueOperationName } from './unique-operation-name/index.js';

export const rules = {
  ...GRAPHQL_JS_VALIDATIONS,
  alphabetize,
  'description-style': descriptionStyle,
  'input-name': inputName,
  'lone-executable-definition': loneExecutableDefinition,
  'match-document-filename': matchDocumentFilename,
  'naming-convention': namingConvention,
  'no-anonymous-operations': noAnonymousOperations,
  'no-deprecated': noDeprecated,
  'no-duplicate-fields': noDuplicateFields,
  'no-hashtag-description': noHashtagDescription,
  'no-one-place-fragments': noOnePlaceFragments,
  'no-root-type': noRootType,
  'no-scalar-result-type-on-mutation': noScalarResultTypeOnMutation,
  'no-typename-prefix': noTypenamePrefix,
  'no-unreachable-types': noUnreachableTypes,
  'no-unused-fields': noUnusedFields,
  'relay-arguments': relayArguments,
  'relay-connection-types': relayConnectionTypes,
  'relay-edge-types': relayEdgeTypes,
  'relay-page-info': relayPageInfo,
  'require-deprecation-date': requireDeprecationDate,
  'require-deprecation-reason': requireDeprecationReason,
  'require-description': requireDescription,
  'require-field-of-type-query-in-mutation-result': requireFieldOfTypeQueryInMutationResult,
  'require-import-fragment': requireImportFragment,
  'require-nullable-fields-with-oneof': requireNullableFieldsWithOneof,
  'require-nullable-result-in-root': requireNullableResultInRoot,
  'require-selections': requireSelections,
  'require-type-pattern-with-oneof': requireTypePatternWithOneof,
  'selection-set-depth': selectionSetDepth,
  'strict-id-in-types': strictIdInTypes,
  'unique-enum-value-names': uniqueEnumValueNames,
  'unique-fragment-name': uniqueFragmentName,
  'unique-operation-name': uniqueOperationName,
};
