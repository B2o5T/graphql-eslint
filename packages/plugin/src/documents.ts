import { resolve } from 'path';
import {
  FragmentDefinitionNode,
  Kind,
  OperationDefinitionNode,
  SelectionSetNode,
  visit,
  OperationTypeNode,
} from 'graphql';
import { Source } from '@graphql-tools/utils';
import { GraphQLProjectConfig } from 'graphql-config';
import debugFactory from 'debug';
import fg from 'fast-glob';
import { logger } from './utils.js';
import { Pointer } from './types.js';
import { ModuleCache } from './cache.js';

export type FragmentSource = { filePath: string; document: FragmentDefinitionNode };
export type OperationSource = { filePath: string; document: OperationDefinitionNode };

const debug = debugFactory('graphql-eslint:operations');

export type SiblingOperations = {
  available: boolean;
  getFragment(fragmentName: string): FragmentSource[];
  getFragments(): FragmentSource[];
  getFragmentByType(typeName: string): FragmentSource[];
  getFragmentsInUse(
    baseOperation: OperationDefinitionNode | FragmentDefinitionNode | SelectionSetNode,
    recursive?: boolean,
  ): FragmentDefinitionNode[];
  getOperation(operationName: string): OperationSource[];
  getOperations(): OperationSource[];
  getOperationByType(operationType: OperationTypeNode): OperationSource[];
};

const handleVirtualPath = (documents: Source[]): Source[] => {
  const filepathMap: Record<string, number> = Object.create(null);

  return documents.map(source => {
    const location = source.location!;
    if (['.gql', '.graphql'].some(extension => location.endsWith(extension))) {
      return source;
    }
    filepathMap[location] ??= -1;
    const index = (filepathMap[location] += 1);
    return {
      ...source,
      location: resolve(location, `${index}_document.graphql`),
    };
  });
};

const operationsCache = new ModuleCache<Source[]>();
const siblingOperationsCache = new Map<Source[], SiblingOperations>();

const getSiblings = (project: GraphQLProjectConfig): Source[] => {
  const documentsKey = project.documents;
  if (!documentsKey) {
    return [];
  }

  let siblings = operationsCache.get(documentsKey);

  if (!siblings) {
    debug('Loading operations from %o', project.documents);
    const documents = project.loadDocumentsSync(project.documents, {
      skipGraphQLImport: true,
      pluckConfig: project.extensions.pluckConfig,
    });
    if (debug.enabled) {
      debug('Loaded %d operations', documents.length);
      const operationsPaths = fg.sync(project.documents as Pointer, { absolute: true });
      debug('Operations pointers %O', operationsPaths);
    }
    siblings = handleVirtualPath(documents);
    operationsCache.set(documentsKey, siblings);
  }

  return siblings;
};

export function getDocuments(project: GraphQLProjectConfig): SiblingOperations {
  const siblings = getSiblings(project);

  if (siblings.length === 0) {
    let printed = false;

    const noopWarn = () => {
      if (!printed) {
        logger.warn(
          'getSiblingOperations was called without any operations. Make sure to set "parserOptions.operations" to make this feature available!',
        );
        printed = true;
      }
      return [];
    };

    return {
      available: false,
      getFragment: noopWarn,
      getFragments: noopWarn,
      getFragmentByType: noopWarn,
      getFragmentsInUse: noopWarn,
      getOperation: noopWarn,
      getOperations: noopWarn,
      getOperationByType: noopWarn,
    };
  }

  // Since the siblings array is cached, we can use it as cache key.
  // We should get the same array reference each time we get
  // to this point for the same graphql project
  const value = siblingOperationsCache.get(siblings);

  if (value) {
    return value;
  }

  let fragmentsCache: FragmentSource[] | null = null;

  const getFragments = (): FragmentSource[] => {
    if (fragmentsCache === null) {
      const result: FragmentSource[] = [];

      for (const source of siblings) {
        for (const definition of source.document?.definitions || []) {
          if (definition.kind === Kind.FRAGMENT_DEFINITION) {
            result.push({
              filePath: source.location!,
              document: definition,
            });
          }
        }
      }
      fragmentsCache = result;
    }
    return fragmentsCache;
  };

  let cachedOperations: OperationSource[] | null = null;

  const getOperations = (): OperationSource[] => {
    if (cachedOperations === null) {
      const result: OperationSource[] = [];

      for (const source of siblings) {
        for (const definition of source.document?.definitions || []) {
          if (definition.kind === Kind.OPERATION_DEFINITION) {
            result.push({
              filePath: source.location!,
              document: definition,
            });
          }
        }
      }
      cachedOperations = result;
    }
    return cachedOperations;
  };

  const getFragment = (name: string) => getFragments().filter(f => f.document.name?.value === name);

  const collectFragments = (
    selectable: SelectionSetNode | OperationDefinitionNode | FragmentDefinitionNode,
    recursive: boolean,
    collected = new Map<string, FragmentDefinitionNode>(),
  ) => {
    visit(selectable, {
      FragmentSpread(spread) {
        const fragmentName = spread.name.value;
        const [fragment] = getFragment(fragmentName);

        if (!fragment) {
          logger.warn(
            `Unable to locate fragment named "${fragmentName}", please make sure it's loaded using "parserOptions.operations"`,
          );
          return;
        }
        if (!collected.has(fragmentName)) {
          collected.set(fragmentName, fragment.document);
          if (recursive) {
            collectFragments(fragment.document, recursive, collected);
          }
        }
      },
    });
    return collected;
  };

  const siblingOperations: SiblingOperations = {
    available: true,
    getFragment,
    getFragments,
    getFragmentByType: typeName =>
      getFragments().filter(f => f.document.typeCondition?.name?.value === typeName),
    getFragmentsInUse: (selectable, recursive = true) =>
      Array.from(collectFragments(selectable, recursive).values()),
    getOperation: name => getOperations().filter(o => o.document.name?.value === name),
    getOperations,
    getOperationByType: type => getOperations().filter(o => o.document.operation === type),
  };

  siblingOperationsCache.set(siblings, siblingOperations);
  return siblingOperations;
}
