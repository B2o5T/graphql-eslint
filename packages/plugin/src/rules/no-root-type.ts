import { NameNode } from 'graphql';
import { ARRAY_DEFAULT_OPTIONS, requireGraphQLSchemaFromContext } from '../utils';
import { GraphQLESLintRule } from '../types';
import { GraphQLESTreeNode } from '../estree-parser';

const ROOT_TYPES: ('mutation' | 'subscription')[] = ['mutation', 'subscription'];

type NoRootTypeConfig = { disallow: typeof ROOT_TYPES };

const rule: GraphQLESLintRule<[NoRootTypeConfig]> = {
  meta: {
    type: 'suggestion',
    hasSuggestions: true,
    docs: {
      category: 'Schema',
      description: 'Disallow using root types `mutation` and/or `subscription`.',
      url: 'https://github.com/B2o5T/graphql-eslint/blob/master/docs/rules/no-root-type.md',
      requiresSchema: true,
      isDisabledForAllConfig: true,
      examples: [
        {
          title: 'Incorrect',
          usage: [{ disallow: ['mutation', 'subscription'] }],
          code: /* GraphQL */ `
            type Mutation {
              createUser(input: CreateUserInput!): User!
            }
          `,
        },
        {
          title: 'Correct',
          usage: [{ disallow: ['mutation', 'subscription'] }],
          code: /* GraphQL */ `
            type Query {
              users: [User!]!
            }
          `,
        },
      ],
    },
    schema: {
      type: 'array',
      minItems: 1,
      maxItems: 1,
      items: {
        type: 'object',
        additionalProperties: false,
        required: ['disallow'],
        properties: {
          disallow: {
            ...ARRAY_DEFAULT_OPTIONS,
            items: {
              enum: ROOT_TYPES,
            },
          },
        },
      },
    },
  },
  create(context) {
    const schema = requireGraphQLSchemaFromContext('no-root-type', context);
    const disallow = new Set(context.options[0].disallow);

    const rootTypeNames = [
      disallow.has('mutation') && schema.getMutationType(),
      disallow.has('subscription') && schema.getSubscriptionType(),
    ]
      .filter(Boolean)
      .map(type => type.name);

    if (rootTypeNames.length === 0) {
      return {};
    }

    const selector = [
      `:matches(ObjectTypeDefinition, ObjectTypeExtension)`,
      '>',
      `Name[value=/^(${rootTypeNames.join('|')})$/]`,
    ].join(' ');

    return {
      [selector](node: GraphQLESTreeNode<NameNode>) {
        const typeName = node.value;
        context.report({
          node,
          message: `Root type \`${typeName}\` is forbidden.`,
          suggest: [
            {
              desc: `Remove \`${typeName}\` type`,
              fix: fixer => fixer.remove(node.parent as any),
            },
          ],
        });
      },
    };
  },
};

export default rule;
