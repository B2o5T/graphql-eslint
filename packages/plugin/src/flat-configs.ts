import { parseForESLint } from './parser';
import { rules } from './rules';
import { configs } from './configs';

const baseOptions = {
  parser: { parseForESLint },
  plugins: {
    '@graphql-eslint': { rules },
  },
};

export const flatConfigs = {
  'operations-all': {
    ...baseOptions,
    rules: {
      ...configs['operations-recommended'].rules,
      ...configs['operations-all'].rules,
    },
  },
  'operations-recommended': {
    ...baseOptions,
    rules: configs['operations-recommended'].rules,
  },
  relay: {
    ...baseOptions,
    rules: configs.relay.rules,
  },
  'schema-all': {
    ...baseOptions,
    rules: {
      ...configs['schema-recommended'].rules,
      ...configs['schema-all'].rules,
    },
  },
  'schema-recommended': {
    ...baseOptions,
    rules: configs['schema-recommended'].rules,
  },
};
