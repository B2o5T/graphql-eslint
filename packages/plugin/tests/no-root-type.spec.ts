import { GraphQLRuleTester, ParserOptions } from '../src';
import { rule } from '../src/rules/no-root-type';

const useSchema = (
  code: string,
  schema = '',
): { code: string; parserOptions: Pick<ParserOptions, 'schema'> } => ({
  code,
  parserOptions: {
    schema: schema + code,
  },
});

const ruleTester = new GraphQLRuleTester();

ruleTester.runGraphQLTests('no-root-type', rule, {
  valid: [
    {
      ...useSchema('type Query'),
      options: [{ disallow: ['mutation', 'subscription'] }],
    },
  ],
  invalid: [
    {
      ...useSchema('type Mutation'),
      name: 'disallow mutation',
      options: [{ disallow: ['mutation'] }],
      errors: [{ message: 'Root type `Mutation` is forbidden.' }],
    },
    {
      ...useSchema('type Subscription'),
      name: 'disallow subscription',
      options: [{ disallow: ['subscription'] }],
      errors: [{ message: 'Root type `Subscription` is forbidden.' }],
    },
    {
      ...useSchema('extend type Mutation { foo: ID }', 'type Mutation'),
      name: 'disallow with extend',
      options: [{ disallow: ['mutation'] }],
      errors: [{ message: 'Root type `Mutation` is forbidden.' }],
    },
    {
      ...useSchema('type MyMutation', 'schema { mutation: MyMutation }'),
      name: 'disallow when root type name is renamed',
      options: [{ disallow: ['mutation'] }],
      errors: [{ message: 'Root type `MyMutation` is forbidden.' }],
    },
  ],
});
