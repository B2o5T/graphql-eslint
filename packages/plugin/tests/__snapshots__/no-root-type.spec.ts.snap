// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
> 1 | type Mutation
    |      ^^^^^^^^ Root type "Mutation" is forbidden
`;

exports[` 2`] = `
> 1 | type Subscription
    |      ^^^^^^^^^^^^ Root type "Subscription" is forbidden
`;

exports[` 3`] = `
> 1 | extend type Mutation { foo: ID }
    |             ^^^^^^^^ Root type "Mutation" is forbidden
`;

exports[` 4`] = `
> 1 | type MyMutation
    |      ^^^^^^^^^^ Root type "MyMutation" is forbidden
`;
